### PR TITLE
Add polygon selection

### PR DESCRIPTION
**Pull request by jfoster17**
# Pull Request Template

## Description

This adds a polygon/lasso-type selection tool to the image and scatter viewers in glue-jupyter using the MouseInteraction class provided in bqplot_image_gl. Creating subsets from arbitrary regions is a top priority for Alyssa before we show an alpha version of glue-jupyter-lab to the world. 

For similarity with other selection tools it would have been nice to get the data back from the LassoSelector (as in [this stalled PR](https://github.com/bqplot/bqplot/pull/674)), but getting that finished and into a released version of bqplot was probably going to take too long.

I hope that the new testing infrastructure can verify that adding a new default tool to the basic viewers does not mess things up for jdaviz.

### CODE REVIEW

**Comment by jfoster17** [comment]
Tagging @astrofrog and @maartenbreddels for their awareness here

**Comment by jfoster17** [comment]
Not sure what the CircleCI error here means...

File: glue_jupyter/bqplot/common/tools.py [hunk 1]
```diff
@@ -178,6 +178,73 @@ def activate(self):
         super().activate()
 
 
+@viewer_tool
+class BqplotPolygonMode(BqplotSelectionTool):
+
+    icon = os.path.join(ICONS_DIR, 'glue_lasso')
```
Comment (astrofrog) [inline glue_jupyter/bqplot/common/tools.py:184]:
```suggestion
    icon = 'glue_lasso'
```

should work as it's a standard icon in glue-core

**Review by astrofrog — CHANGES_REQUESTED** [review]
Aside from a small comment about the icon below, I couldn't get this to work - for some reason the drag events are just not triggered. If I add 'click' to the events to listen to, then I can check that on_msg gets called when I click, but nothing happens when I click and drag. I'm using MacOS X and I see this on both Safari and Google Chrome.

**Comment by astrofrog** [comment]
The CircleCI error should be fixed by https://github.com/glue-viz/glue-jupyter/pull/372

File: glue_jupyter/bqplot/common/tools.py (hunk 1 above)
Comment (jfoster17) [reply glue_jupyter/bqplot/common/tools.py]: Can you confirm if you were trying this in Jupyter Lab or Jupyter Notebook? I was developing in Jupyter Lab, in Notebook I get what seems like your problem. Just want to confirm that it's the same thing...

Comment (astrofrog) [reply glue_jupyter/bqplot/common/tools.py]: I think I was using notebook

**Comment by astrofrog** [comment]
I just tried this out in Jupyter lab and it works nicely! One small difference compared to other tools is that when the selection is being drawn, it is bright yellow while drawing is active, whereas for other selection tools it is a translucent grey. Would it be possible to make this new tool match the behavior of the existing ones?

**Comment by jfoster17** [comment]
> I just tried this out in Jupyter lab and it works nicely! One small difference compared to other tools is that when the selection is being drawn, it is bright yellow while drawing is active, whereas for other selection tools it is a translucent grey. Would it be possible to make this new tool match the behavior of the existing ones?

Done.

**Comment by jfoster17** [comment]
Do you get any error when trying to use this polygon tool in Notebook? I get:

```
[E 11:07:22.922 NotebookApp] Uncaught exception, closing connection.
    Traceback (most recent call last):
      File "/Users/jfoster/mambaforge/envs/bqplot-polygon-selection/lib/python3.11/site-packages/tornado/iostream.py", line 695, in _handle_events
        self._handle_write()
      File "/Users/jfoster/mambaforge/envs/bqplot-polygon-selection/lib/python3.11/site-packages/tornado/iostream.py", line 965, in _handle_write
        self._write_buffer.advance(num_bytes)
      File "/Users/jfoster/mambaforge/envs/bqplot-polygon-selection/lib/python3.11/site-packages/tornado/iostream.py", line 182, in advance
        assert 0 < size <= self._size
               ^^^^^^^^^^^^^^^^^^^^^^
    AssertionError
Exception in callback None()
```

which looks suspiciously similar to the iostream messages reported [here](https://github.com/jupyter/notebook/issues/6721).

**Comment by astrofrog** [comment]
No, nothing happened before when I tried to click on the figure.

**Comment by astrofrog** [comment]
@jfoster17 - when I use notebook I don't see any errors - either in the browser console or in the terminal where I started the notebook from. Clicking and dragging just doesn't do anything.

**Comment by Carifio24** [comment]
I'm seeing the same thing - works great in lab, but no errors and no functionality in the notebook. I found something that, at least for me, makes it work in both environments, but maybe @astrofrog has a bit more insight into _why_ this would be the case.

So I think the issue might be the `MouseInteraction` that's set as the tool's `interact` member (maybe it's related to the comment [here](https://github.com/glue-viz/bqplot-image-gl/blob/master/js/lib/MouseInteraction.js#L186)?). As it is now, for me all drag events in the notebook are completely silenced with this tool active - it seems that this interaction is eating them up somehow. But if I ditch this interaction (set `self.interact = None`), this works fine for me in both the notebook and lab. (Why this isn't an issue in lab, I have no clue).

**Comment by jfoster17** [comment]
Fantastic! Thanks for this breakthrough Jon. Now there are just a few more things to clean up...

**Comment by jfoster17** [comment]
I think this is good-to-go. I rebased to get the CircleCI fix. @astrofrog -- would you like to do another review?

**Review by astrofrog — APPROVED** [review]
Works great in both the scatter and image viewer, thanks! 🎉

**Pull request merged**